File: Scala/string_algorithms/src/main/scala/suffix_trees/GeneralizedPatternMatching.scala
package suffix_trees

object GeneralizedPatternMatching {
  import scala.annotation.tailrec

  private type Node = Int

  final case class EdgeEndpoint(tip: Node, label: Char)

  private val stringTerminator: Char = '$'

  class Trie(words: List[String]) {
    import Trie.getNeighbourWithGivenLabel

    private val nodeCounter: Iterator[Int] = Iterator.from(0)
    private val root: Int = nodeCounter.next()
    private val adjacencyList: Map[Node, List[EdgeEndpoint]] = addWordsToTrie(this.words)

    def patternStarts(text: List[Char]): Boolean = {
      @tailrec
      def loop(ls: List[Char], currentNode: Node): Boolean = ls match {
        case Nil => false
        case letter :: lss => getNeighbourWithGivenLabel(adjacencyList, currentNode, letter) match {
          case None => false
          case Some(nextNode) => if (hasLeafChild(nextNode)) true else loop(lss, nextNode)
        }
      }

      loop(text, root)
    }

    private def hasLeafChild(node: Node): Boolean =
      adjacencyList.getOrElse(node, Nil).exists{ case EdgeEndpoint(_, label) => label == stringTerminator }

    private def addWordsToTrie(words: List[String]): Map[Node, List[EdgeEndpoint]] = {
      @tailrec
      def loop(ws: List[String], trie: Map[Node, List[EdgeEndpoint]]): Map[Node, List[EdgeEndpoint]] = ws match {
        case Nil => trie
        case w :: wss =>
          val updatedTrie: Map[Node, List[EdgeEndpoint]] = addWord(trie, w)
          loop(wss, updatedTrie)
      }

      loop(words, Map())
    }

    private def addWord(trie: Map[Node, List[EdgeEndpoint]], w: String): Map[Node, List[EdgeEndpoint]] = {
      @tailrec
      def loop(ls: List[Char],
               acc: Map[Node, List[EdgeEndpoint]],
               currentNode: Node): Map[Node, List[EdgeEndpoint]] = ls match {
        case Nil => acc
        case letter :: lss => getNeighbourWithGivenLabel(acc, currentNode, letter) match {
          case None =>
            val nextNode: Node = nodeCounter.next()
            val updatedTrie: Map[Node, List[EdgeEndpoint]] =
              acc.updated(currentNode, EdgeEndpoint(nextNode, letter) :: acc.getOrElse(currentNode, Nil))
            loop(lss, updatedTrie, nextNode)
          case Some(node) => loop(lss, acc, node)
        }
      }

      loop(w.toList, trie, root)
    }
  }

  object Trie {
    private def getNeighbourWithGivenLabel(trie: Map[Node, List[EdgeEndpoint]],
                                           currentNode: Node,
                                           letter: Char): Option[Node] = {
      trie
        .get(currentNode)
        .flatMap(_.find{ case EdgeEndpoint(_, label) => label == letter })
        .map(_.tip)
    }
  }

  private[suffix_trees] def appendTerminator(string: String): String = string + stringTerminator.toString

  def multiplePatternMatching(text: String, patterns: List[String]): List[Int] = {
    val trie = new Trie(patterns)
    @tailrec
    def loop(indices: List[Int], text: List[Char], ix: Int): List[Int] = text match {
      case Nil => indices.reverse
      case _ :: rest =>
        val updatedIndices: List[Int] = if (trie.patternStarts(text)) ix :: indices else indices
        loop(updatedIndices, rest, ix + 1)
    }

    loop(Nil, text.toList, 0)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val text: String = reader.next()
    val nrPatterns: Int = reader.next().toInt
    val patterns: List[String] = reader.take(nrPatterns).map(appendTerminator).toList
    val result: List[Int] = multiplePatternMatching(text, patterns)
    println(result.mkString(" "))
  }
}

File: OCaml/string_algorithms/suffix_tries/generalized_pattern_matching.ml
let string_terminator = '$'


module IntMap = Map.Make(
    struct
        type t = int
        let compare = Pervasives.compare
    end)

module Trie : sig
        type t
        val create : string list -> t
        val pattern_starts : char list -> t -> bool
    end = struct
        open Batteries

        type node = int
        type edge_endpoint = { tip : node; label : char }
        type t = edge_endpoint list IntMap.t

        let root = 0

        let get_neighbour_with_given_label trie current_node letter =
            Option.map (function { tip } -> tip ) @@
                Option.bind (IntMap.find_opt current_node trie) (List.find_opt (function { label } -> label = letter ))

        let has_leaf_child node trie = match IntMap.find_opt node trie with
            | None -> false
            | Some children -> List.exists (function { label } -> label = string_terminator) children

        let add_word trie w node_counter =
            let rec loop acc current_node = function
                | [] -> acc
                | letter :: lss -> match (get_neighbour_with_given_label acc current_node letter) with
                    | None ->
                        let next_node = !node_counter in
                        node_counter := !node_counter + 1;
                        let children = match IntMap.find_opt current_node acc with
                            | None -> []
                            | Some values -> values in
                        let updated_trie = IntMap.add current_node ({ tip = next_node; label = letter } :: children) acc
                        in loop updated_trie next_node lss
                    | Some node -> loop acc node lss
            in loop trie root (String.to_list w)

        let create words =
            let node_counter = ref 1 in
            let rec loop trie = function
                | [] -> trie
                | w :: wss ->
                    let updated_trie = add_word trie w node_counter
                    in loop updated_trie wss
            in loop IntMap.empty words

        let pattern_starts text trie =
            let rec loop current_node = function
                | [] -> false
                | letter :: lss -> match (get_neighbour_with_given_label trie current_node letter) with
                    | None -> false
                    | Some next_node -> if has_leaf_child next_node trie then true else loop next_node lss
            in loop root text
    end


let append_terminator string = string ^ Char.escaped string_terminator


let multiple_pattern_matching text patterns =
    let trie = Trie.create patterns in
    let rec loop indices suffix ix = match suffix with
        | [] -> List.rev indices
        | _ :: rest ->
            let updated_indices = if Trie.pattern_starts suffix trie then ix :: indices else indices in
            loop updated_indices rest (ix + 1) in
    let char_list = let open Batteries in String.to_list text in
    loop [] char_list 0


let read_patterns n =
    let rec loop acc k =
        if k = n then List.rev acc
        else loop (read_line() :: acc) (k + 1)
    in loop [] 0


let () =
    let text = read_line() in
    let number_of_patterns = read_int() in
    let patterns = number_of_patterns |> read_patterns |> List.map append_terminator in
    let result = multiple_pattern_matching text patterns in
    List.iter (function ix -> print_int ix; print_string " ") result;
    print_newline()
